File: exercises/practice/knapsack/knapsack.zig
const std = @import("std");
const mem = std.mem;

pub const Item = struct {
    // This struct, as well as its fields and init method, needs to be implemented.

    pub fn init(weight: usize, value: usize) Item {
        _ = weight;
        _ = value;
        @compileError("please implement the init method");
    }
};

pub fn maximumValue(allocator: mem.Allocator, maximumWeight: usize, items: []const Item) !usize {
    _ = allocator;
    _ = maximumWeight;
    _ = items;
    @compileError("please implement the maximumValue function");
}

File: generators/exercises/knapsack.py
HEADER = "const Item = knapsack.Item;\n"


def gen_case(case):
    input = case["input"]
    items = input["items"]
    maximum_weight = input["maximumWeight"]
    expected = case["expected"]

    lines = [f"    const expected: usize = {expected};\n"]
    lines.append(f"    const items: [{len(items)}]Item = .{{\n")
    for item in items:
        lines.append(f"        Item.init({item['weight']}, {item['value']}),\n")
    lines.append("    };\n")
    lines.append(
        f"    const actual = try knapsack.maximumValue(testing.allocator, {maximum_weight}, &items);\n"
    )
    lines.append("    try testing.expectEqual(expected, actual);\n")
    return "".join(lines)
